cli/internal/specs/v0/spec_reader.go

```diff
@@ -0,0 +1,246 @@
+		}
+		doc, err = expandEnv(doc)
+		if err != nil {
+			return fmt.Errorf("failed to expand environment variable in file %s (section %d): %w", path, i+1, err)
+		}
+		var s Spec
+		if err := SpecUnmarshalYamlStrict(doc, &s); err != nil {
+			return fmt.Errorf("failed to unmarshal file %s: %w", path, err)
+		}
+		switch s.Kind {
+		case KindSource:
+			source := s.Spec.(*Source)
+			if r.sourcesMap[source.Name] != nil {
+				return fmt.Errorf("duplicate source name %s", source.Name)
+			}
+			source.SetDefaults()
+			if err := source.Validate(); err != nil {
+				return fmt.Errorf("failed to validate source %s: %w", source.Name, err)
+			}
+			r.sourcesMap[source.Name] = source
+			r.Sources = append(r.Sources, source)
+		case KindDestination:
+			destination := s.Spec.(*Destination)
+			if r.destinationsMap[destination.Name] != nil {
+				return fmt.Errorf("duplicate destination name %s", destination.Name)
+			}
+			// We set the default value to 0, so it can be overridden later by plugins' defaults
+			destination.SetDefaults(0, 0)
+			if err := destination.Validate(); err != nil {
+				return fmt.Errorf("failed to validate destination %s: %w", destination.Name, err)
+			}
+			r.destinationsMap[destination.Name] = destination
+			r.Destinations = append(r.Destinations, destination)
+		default:
+			return fmt.Errorf("unknown kind %s", s.Kind)
+		}
+	}
+	return nil
+}
+
+func (r *SpecReader) loadSpecsFromDir(path string) error {
+	files, err := os.ReadDir(path)
+	if err != nil {
+		return fmt.Errorf("failed to read directory %s: %w", path, err)
+	}
+	for _, file := range files {
+		if !file.IsDir() && !strings.HasPrefix(file.Name(), ".") &&
+			(strings.HasSuffix(file.Name(), ".yml") || strings.HasSuffix(file.Name(), ".yaml")) {
+			if err := r.loadSpecsFromFile(filepath.Join(path, file.Name())); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (r *SpecReader) validate() error {
+	if len(r.Sources) == 0 {
+		return fmt.Errorf("expecting at least one source")
+	}
+	if len(r.Destinations) == 0 {
+		return fmt.Errorf("expecting at least one destination")
+	}
+
+	// here we check if source with different versions use the same destination and error out if yes
+	var destinationSourceMap = make(map[string]string)
+	for _, source := range r.Sources {
+		for _, destination := range source.Destinations {
+			if r.destinationsMap[destination] == nil {
+				return fmt.Errorf("source %s references unknown destination %s", source.Name, destination)
+			}
+			destinationToSourceKey := fmt.Sprintf("%s-%s", destination, source.Path)
+			if destinationSourceMap[destinationToSourceKey] == "" {
+				destinationSourceMap[destinationToSourceKey] = source.Path + "@" + source.Version
+			} else if destinationSourceMap[destinationToSourceKey] != source.Path+"@"+source.Version {
+				return fmt.Errorf("destination %s is used by multiple sources %s with different versions", destination, source.Path)
+			}
+		}
+	}
```

Comment:
Is this validation step a new addition? I don't quite understand what it's doing / why we can't allow different source versions can't use the same destination

Reply:
Looks like it's not new in any case; but I'm still curious why it exists 😄